drivers:adc:ad7193: Code Updated

pmallick-adi opened the pull request:
* Added other device parameters to device descriptor and init param
* Added functions to configure these additional parameters.
* The channel select function can now enable multiple channels.
* The single read function now samples data with rate defined in the descriptor.
* Update the init function to reset and initialize the device with the init param.
*  All functions now return error status.
* Added Astyle Fix.

Signed-off-by: Pratyush Mallick <Pratyush.Mallick@analog.com>

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -211,34 +223,37 @@ void ad7193_reset(struct ad7193_dev *dev)
+	uint32_t new_mode = 0x0;
+
+	old_mode  = ad7193_get_register_value(dev,
+					      AD7193_REG_MODE,
```
nit: the nr of lines here can be reduced. This applies for all other `ad7193_get_register_value`/`ad7193_set_register_value` in this commit.

pmallick-adi replied:
Sure, will fix them

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -96,6 +93,21 @@ int8_t ad7193_init(struct ad7193_dev **device,
 	}
 
+	/* Reset */
+	AD7193_CS_LOW;
```
since we update de driver, I'd remove these definitions: https://github.com/analogdevicesinc/no-OS/blob/cbd8405892eeadc4383a6d8ddc3644fb9a6f4743/drivers/adc/ad7193/ad7193.h#L58-L65 in a separate commit and use gpio getters/setters inline. But that's your call 😄

pmallick-adi replied:
Will, update this as well.

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -319,6 +334,163 @@ void ad7193_calibrate(struct ad7193_dev *dev,
+			 3,
+			 1);
+
+	old_reg_value &= ~(AD7193_CONF_PSEUDO);
```
Why not adding a generic `ad7193_update_register value` similar to:
https://github.com/analogdevicesinc/no-OS/blob/a2f59796c297b03cf5f36a59d4d4d441b95b5230/drivers/adc/ad463x/ad463x.c#L111-L127

Most of these functions that your are adding seem to update specific bits of a register.

pmallick-adi replied:
Will implement this.

adrimbarean commented on drivers/adc/ad7193/ad7193.h:
```diff
@@ -175,6 +175,9 @@
 #define AD7193_GPOCON_P0DAT     (1 << 0) // P0 state
 
+/* Channel Mask */
+#define AD7193_CH_MASK(x)		(1 << x)
```
I think we could use the [BIT](https://github.com/analogdevicesinc/no-OS/blob/master/include/util.h) macro here.

pmallick-adi replied:
Will use it here.

adrimbarean commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -211,34 +223,37 @@ void ad7193_reset(struct ad7193_dev *dev)
 *******************************************************************************/
-void ad7193_set_power(struct ad7193_dev *dev,
-		      uint8_t pwr_mode)
+void ad7193_set_operating_mode(struct ad7193_dev *dev,
```
Nitpick: not necessary, but I'd take this chance to implement status return on these functions if it's not too much hassle. We should always have feedback on operations we do.

pmallick-adi replied:
Do you want me update all the function with status return type ?

amiclaus reviewed (commented):
> Do you want me update all the function with status return type ?
>
@pmallick-adi this can be done in a separate commit, but since it is not the main purpose of this PR, I think it is your decision if you want to implement it or not.

pmallick-adi commented:
> > Do you want me update all the function with status return type ?
> 
> @pmallick-adi this can be done in a separate commit, but since it is not the main purpose of this PR, I think it is your decision if you want to implement it or not.

Thanks for the clarification. I'll update this PR with the suggested changes except for updating the return type. I'll make a separate PR for this.

pmallick-adi commented:
> > Do you want me update all the function with status return type ?
> 
> @pmallick-adi this can be done in a separate commit, but since it is not the main purpose of this PR, I think it is your decision if you want to implement it or not.

This PR also includes updates for returning error status from each function.

amiclaus reviewed (commented):
Please add a changelog, stating all the changes that were made with your latest push.

Pratyush-Mallick commented:
Changelog: 

* Added the masked_set_register function and updated functions to make use of the same. 
* Updated functions to return error status. 
* Update the initialization function. 
* Updated functions to make use of use gpio getters/setters inline.
* Added Astyle fix. 
* Shorted variable names.

pmallick-adi commented:
Changelog: 

* Added the masked_set_register function and updated functions to make use of the same. 
* Updated functions to return error status. 
* Update the initialization function. 
* Updated functions to make use of use gpio getters/setters inline.
* Added Astyle fix. 
* Shorted variable names.

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -329,125 +580,150 @@ void ad7193_calibrate(struct ad7193_dev *dev,
+		return ret;
+
+	// CS is not modified.
+	ad7193_set_register_value(dev, AD7193_REG_MODE, command, 3, 0);
```
missing return check

pmallick-adi replied:
Will add it.

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -53,236 +54,356 @@
+	old_reg_data &= ~mask;
+	new_reg_data = old_reg_data | reg_data;
+
+	ret = ad7193_set_register_value(dev, reg_addr, new_reg_data, bytes, modify_cs);
```
you can return directly here

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -53,236 +54,356 @@
+		return -EINVAL;
+	}
+
+	ret = ad7193_set_masked_register_value(dev, AD7193_REG_CONF,
```
return directly

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -292,31 +413,161 @@ void ad7193_channel_select(struct ad7193_dev *dev,
+	if (ret != SUCCESS)
+		return ret;
+
+	ret = gpio_set_value(dev->gpio_cs, GPIO_HIGH);
```
return directly.

amiclaus reviewed (commented):
Some comments on my side regarding the new version of the patch.
Also, you might want to somehow split these commits, one for adding return check, one for the functionalities that were added.

pmallick-adi commented:
> Some comments on my side regarding the new version of the patch. Also, you might want to somehow split these commits, one for adding return check, one for the functionalities that were added.

Sure, will split commits.

pmallick-adi commented:
Changelog:

Just separated out the pervious changes in two commits: 
* Return error status. 
* Functional Updates.

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -463,33 +603,32 @@ int32_t ad7193_range_setup(struct ad7193_dev *dev,
 int32_t ad7193_single_conversion(struct ad7193_dev *dev, uint32_t *reg_data)
 {
 	uint32_t command = 0x0;
+
```
nit: extra line.

pmallick-adi replied:
Removed it.

amiclaus reviewed (approved):
one more nit, rest looks good (commits structure can be polished further, but if no one else has complaints, i'm fine like this)

pmallick-adi commented:
Change Log: 

Resolved the merge conflicts due to change in path to the include header files.

buha commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -42,6 +42,7 @@
 /******************************************************************************/
 #include <stdlib.h>
 #include "ad7193.h"    // AD7193 definitions.
+#include "error.h"
```
yeah sorry for the trouble @pmallick-adi, we moved the include files around and all ongoing PRs now have to modify these paths to no-os/error.h and so on

adrimbarean reviewed (approved):
Looks good to me.

pmallick-adi commented:
Can this be merged now?

buha commented:
Looks good, a few remarks only:

Commits 1 & 4 should be merged
Also, the merged commits 1 & 4 could probably be also merged in one of the remaining commits, whichever you find suitable

We've decided a while ago to use bare `int` as a return type instead of int32_t, int16_t etc. because `int` will always have the best width for the actual processor code is running on. It would be great if you could also refactor this (change all functions that return a ret status to int, change all `int32_t ret;` occurences to `int ret;`)

pmallick-adi commented:
ChangeLog: 

Merged commits to just 2 commits. 
Updated the return type to int.

pmallick-adi commented:
ChangeLog: 

Merged commits to just 2 commits. 
Updated the return type to int.

amiclaus commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -207,82 +326,76 @@ void ad7193_reset(struct ad7193_dev *dev)
-void ad7193_channel_select(struct ad7193_dev *dev,
-			   uint16_t channel)
+int ad7193_channel_select(struct ad7193_dev *dev,
+			  uint16_t channel)
```
nit: I think here should be `chn_mask` instead of `channel`

pmallick-adi commented:
ChangeLog:

Fixed a function parameter.

pmallick-adi commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -55,149 +56,199 @@
 {
-	uint8_t write_command[5] = {0, 0, 0, 0, 0};
-	uint8_t* data_pointer    = (uint8_t*)&register_value;
+	uint8_t write_command[5] = { 0, 0, 0, 0, 0 };
```
Okay, will update this.

pmallick-adi commented on drivers/adc/ad7193/ad7193.c:
```diff
@@ -244,14 +301,20 @@ void ad7193_set_power(struct ad7193_dev *dev,
 
-	while (wait)
-		AD7193_RDY_STATE(wait);
+	while (wait) {
```
Will add it here.

pmallick-adi commented:
ChangeLog:

Added a timeout in the ad7193_wait_rdy_go_low() function. 
Added -ENOMEM to the during the memory allocation failure. 
Removed spaced between parenthesis for uint8_t write_command[5] in the ad7193_set_register_value() function.
